gobblin-data-management/src/main/java/org/apache/gobblin/data/management/version/finder/LookbackDateTimeDatasetVersionFinder.java

```diff
@@ -0,0 +1,98 @@
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.data.management.version.finder;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.joda.time.DateTime;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.joda.time.Period;
+import org.joda.time.format.PeriodFormatter;
+import org.joda.time.format.PeriodFormatterBuilder;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.typesafe.config.Config;
+
+import org.apache.gobblin.data.management.version.FileSystemDatasetVersion;
+import org.apache.gobblin.data.management.version.TimestampedDatasetVersion;
+import org.apache.gobblin.dataset.Dataset;
+import org.apache.gobblin.dataset.FileSystemDataset;
+import org.apache.gobblin.util.ConfigUtils;
+
+
+/**
+ * {@link DatasetVersionFinder} that constructs {@link TimestampedDatasetVersion}s without actually checking for existence
+ * of the version path. The version path is constructed by appending the version partition pattern to the dataset root.
+ * The versions are found by looking back a specific period of time and finding unique date partitions between that
+ * time and the current time.
+ */
+public class LookbackDateTimeDatasetVersionFinder extends DateTimeDatasetVersionFinder {
+  public static final String VERSION_PATH_PREFIX = "version.path.prefix";
+  public static final String VERSION_LOOKBACK_PERIOD = "version.lookback.period";
+
+  private final Duration stepDuration;
+  private final Period lookbackPeriod;
+  private final String pathPrefix;
+  private final Instant endTime;
+
+  public LookbackDateTimeDatasetVersionFinder(FileSystem fs, Config config) {
+    this(fs, config, Instant.now());
+  }
+
+  @VisibleForTesting
+  public LookbackDateTimeDatasetVersionFinder(FileSystem fs, Config config, Instant endTime) {
+    super(fs, config);
+    Preconditions.checkArgument(config.hasPath(VERSION_LOOKBACK_PERIOD) , "Missing required property " + VERSION_LOOKBACK_PERIOD);
+    PeriodFormatter periodFormatter =
+        new PeriodFormatterBuilder().appendYears().appendSuffix("y").appendMonths().appendSuffix("M").appendDays()
+            .appendSuffix("d").appendHours().appendSuffix("h").appendMinutes().appendSuffix("m").toFormatter();
+    this.stepDuration = Duration.standardMinutes(1);
+    this.pathPrefix = ConfigUtils.getString(config, VERSION_PATH_PREFIX, "");
+    this.lookbackPeriod = periodFormatter.parsePeriod(config.getString(VERSION_LOOKBACK_PERIOD));
+    this.endTime = endTime;
+  }
+
+  @Override
+  public Class<? extends FileSystemDatasetVersion> versionClass() {
+    return TimestampedDatasetVersion.class;
+  }
+
+  @Override
+  public Collection<TimestampedDatasetVersion> findDatasetVersions(Dataset dataset) throws IOException {
```

Comment:
When using iceberg dataset cleaner (what this is intended for), the version finder will be called from `CleanableIcebergDataset`, which will just send GMCEs with those paths and not make any call to the filesystem.

Even if it's used with actual file retention (`MultiVersionCleanableDatasetBase`), I see that in `FsCleanableHelper.clean` it will end up ignoring paths that don't exist so it wouldn't cause any failure.

Reply:
And how do they handle the increase in version numbers? Presumably this will produce more versions than before in some cases. Do we know the magnitude of this increase and how it will affect downstream systems?

Reply:
If we set the iceberg retention lookback to one day longer than the file retention (which is what my plan is), it means each drop_files GMCE will contain a list of 24 previous partitions instead of about 3. I expect there should be little to no performance impact due to that.

Reply:
FYI for future readers, these 24 partitions are later translated to an expression, so it doesn't really matter how long the partition list is
https://github.com/apache/gobblin/blob/0d69363abc776812a4f0352860a4125c7f28a508/gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java#L667-L674